app/src/main/java/org/bepass/oblivion/PublicIPUtils.java

```diff
@@ -0,0 +1,69 @@
+package org.bepass.oblivion;
+
+import android.os.Handler;
+import android.util.Log;
+
+import org.json.JSONObject;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.net.URL;
+
+public class PublicIPUtils {
+
+    public interface IPDetailsCallback {
+        void onDetailsReceived(IPDetails details);
+    }
+
+    public static String convertStreamToString(InputStream is) {
+        BufferedReader reader = new BufferedReader(new InputStreamReader(is));
+        StringBuilder sb = new StringBuilder();
+
+        String line;
+        try {
+            while ((line = reader.readLine()) != null) {
+                sb.append(line).append('\n');
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        } finally {
+            try {
+                is.close();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+
+        return sb.toString();
+    }
+
+    public static void getIPDetails(int port, IPDetailsCallback callback) {
+        Handler handler = new Handler();
+        new Thread(() -> {
+            IPDetails details = new IPDetails();
+            try {
+                Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress("127.0.0.1", port));
```

Comment:
Use SOCKS type as the local proxy is not an HTTP proxy.

Reply:
Thank you for the helpful response explaining the issue. I suspected there may be problems with the Psiphon handshake timing, which is why I had added a 10 second delay before making the IP call. However, I was still seeing no results so there must be another underlying cause. I appreciate you taking ownership of fixing this problem - your expertise will be invaluable.

Regarding using SOCKS instead of HTTP proxies, I did attempt that previously without success.

And about that API for fetching IP data. There are several online tools available for accomplishing this task, but each one has its own problems (accuracy, rate limits, API key requirement, etc). So considering the issues I mentioned, I chose this one overall. But you are right, when checking across different IPs, I saw 403 errors. So I'll try to find a more proper one shortly.

Reply:
Changing the type won't fix this problem. Just a suggestion to use the correct type. I'm working on psiphon connection state. I will let you know once I achieved the desired result, probably in a few hours. Thanks for cooperating to improve the project <3

Reply:
You're welcome. I'm delighted that I get to contribute to this project alongside you and the team. Thank you also for taking the time to address the Psiphon issue properly - I really appreciate it. I'll be awaiting your updates and look forward to rebasing my code once the connection bug is squared away. Please keep up the excellent work!